Route53: retry single changes in a batch if the batch fails

alfredkrohmer opened the pull request:
If a single change fails during the retry, it will be added to a queue.
In the next iteration, changes from this queue will be submitted after
all other changes.

When submitting single changes, they are always submitted as batches of
changes with the same DNS name to avoid inconsistency between the record
created and the TXT records.

This closes #421.

devkid commented:
This fixes #421.

devkid commented:
/assign @njuettner

devkid commented:
Example log:

```
time="2019-09-25T10:52:53Z" level=info msg="Desired change: CREATE ok.external-dns-test.example.com A [Id: /hostedzone/Z2XXXXXXXXXXK]"
time="2019-09-25T10:52:53Z" level=info msg="Desired change: CREATE .external-dns-test.example.com A [Id: /hostedzone/Z2XXXXXXXXXXK]"
time="2019-09-25T10:52:53Z" level=info msg="Desired change: CREATE ok.external-dns-test.example.com TXT [Id: /hostedzone/Z2XXXXXXXXXXK]"
time="2019-09-25T10:52:53Z" level=info msg="Desired change: CREATE .external-dns-test.example.com TXT [Id: /hostedzone/Z2XXXXXXXXXXK]"
time="2019-09-25T10:52:53Z" level=error msg="Failure in zone example.com. [Id: /hostedzone/Z2XXXXXXXXXXK] when submitting change batch"
time="2019-09-25T10:52:53Z" level=error msg="InvalidChangeBatch: [FATAL problem: DomainLabelEmpty (Domain label is empty) encountered with '.external-dns-test.example.com', FATAL problem: DomainLabelEmpty (Domain label is empty) encountered with '.external-dns-test.example.com']\n\tstatus code: 400, request id: 73ba2c80-45b7-438f-a191-f94386d85ed5"
time="2019-09-25T10:52:53Z" level=error msg="Trying to submit changes one-by-one instead"
time="2019-09-25T10:52:53Z" level=info msg="Desired change: CREATE ok.external-dns-test.example.com A [Id: /hostedzone/Z2XXXXXXXXXXK]"
time="2019-09-25T10:52:53Z" level=info msg="Desired change: CREATE ok.external-dns-test.example.com TXT [Id: /hostedzone/Z2XXXXXXXXXXK]"
time="2019-09-25T10:52:53Z" level=info msg="Change successful"
time="2019-09-25T10:52:53Z" level=info msg="Desired change: CREATE .external-dns-test.example.com A [Id: /hostedzone/Z2XXXXXXXXXXK]"
time="2019-09-25T10:52:53Z" level=info msg="Desired change: CREATE .external-dns-test.example.com TXT [Id: /hostedzone/Z2XXXXXXXXXXK]"
time="2019-09-25T10:52:53Z" level=error msg="Failed submitting change, it will be retried in a separate change batch in the next iteration"
time="2019-09-25T10:52:53Z" level=info msg="2 record(s) in zone example.com. [Id: /hostedzone/Z2XXXXXXXXXXK] were successfully updated"
time="2019-09-25T10:52:54Z" level=error msg="Failed to submit all changes for the following zones: [/hostedzone/Z2XXXXXXXXXXK]"
```

next iteration:
```
time="2019-09-25T10:54:53Z" level=info msg="Desired change: CREATE .external-dns-test.example.com A [Id: /hostedzone/Z2XXXXXXXXXXK]"
time="2019-09-25T10:54:53Z" level=info msg="Desired change: CREATE .external-dns-test.example.com TXT [Id: /hostedzone/Z2XXXXXXXXXXK]"
time="2019-09-25T10:54:53Z" level=error msg="Failure in zone example.com. [Id: /hostedzone/Z2XXXXXXXXXXK] when submitting change batch"
time="2019-09-25T10:54:53Z" level=error msg="InvalidChangeBatch: [FATAL problem: DomainLabelEmpty (Domain label is empty) encountered with '.external-dns-test.example.com', FATAL problem: DomainLabelEmpty (Domain label is empty) encountered with '.external-dns-test.example.com']\n\tstatus code: 400, request id: e7dc3ae0-70e7-4090-9050-5b1a53df4782"
time="2019-09-25T10:54:53Z" level=error msg="Trying to submit changes one-by-one instead"
time="2019-09-25T10:54:53Z" level=info msg="Desired change: CREATE .external-dns-test.example.com A [Id: /hostedzone/Z2XXXXXXXXXXK]"
time="2019-09-25T10:54:53Z" level=info msg="Desired change: CREATE .external-dns-test.example.com TXT [Id: /hostedzone/Z2XXXXXXXXXXK]"
time="2019-09-25T10:54:53Z" level=error msg="Failed submitting change, it will be retried in a separate change batch in the next iteration"
time="2019-09-25T10:54:54Z" level=error msg="Failed to submit all changes for the following zones: [/hostedzone/Z2XXXXXXXXXXK]"
```

szuecs commented on provider/aws.go:
```diff
@@ -380,13 +404,56 @@ func (p *AWSProvider) submitChanges(changes []*route53.Change, zones map[string]
-					failedUpdate = true
+
+					if len(b) > 1 {
+						log.Error("Trying to submit changes one-by-one instead")
```
looks more like a debug log to me, or delete it

devkid replied:
Hm are you sure about this? If there is a transaction failing, in my opinion this should output more verbose logs than debug because it needs a manual intervention to fix it.

szuecs commented on provider/aws.go:
```diff
@@ -380,13 +404,56 @@ func (p *AWSProvider) submitChanges(changes []*route53.Change, zones map[string]
+							}
+							if _, err := p.client.ChangeResourceRecordSets(params); err != nil {
+								failedUpdate = true
+								log.Error("Failed submitting change, it will be retried in a separate change batch in the next iteration")
```
looks more like a debug log to me, or delete it

szuecs commented on provider/aws.go:
```diff
@@ -380,13 +404,56 @@ func (p *AWSProvider) submitChanges(changes []*route53.Change, zones map[string]
+									p.failedChangesQueue[z] = append(p.failedChangesQueue[z], groupedChanges...)
+								}
+							} else {
+								log.Info("Change successful")
```
looks more like a debug log to me, or delete it

szuecs commented:
@devkid Thanks for the PR only small comments about the logs from my side.

linki commented:
@devkid Nice idea to group the smaller changes into transactions with the same hostname. Could you double-check if that also works with the `--txt-prefix` feature?

In order to allow TXT records for CNAMEs which cannot co-exist with the same hostname we map between them using a user provided prefix on the TXT record ([See here](https://github.com/kubernetes-incubator/external-dns/blob/v0.5.17/registry/txt.go#L135)). In that configuration the CNAME and its corresponding TXT don't have the same hostname but rather differ by the fixed prefix.

I think the transaction grouping needs to take this into account in order to have the corresponding records in the same transaction.

devkid commented:
@linki Good catch. Would it make sense to attach some kind of meta information to the TXT records (like: "I belong to _this_ record")? Or is there even some information already?

devkid commented:
@linki The existing `batchChangeSet` function also doesn't take the TXT prefix into consideration (which I suppose is quite bad as this can bring things out of sync if a TXT prefix is used). Also, as far as I can tell from looking at the code, there is currently no meta information or so added to the TXT records that associates them with the records they are "managing".

What would you suggest going forward? Shall we merge this pull request first and tackle the "group batches batches so that TXT records are changed in the same transaction" afterwards or should I come up with a solution in this pull request?

For a solution I though about two options:
1. Add a `BelongsTo` (or similar named) fielt to the `Endpoint` struct that can be set to point to another record in the registry implementation (via pointer or name as string).
2. Add an item to the `Labels` map of the TXT record itself with a well-known key that contains the name of the record it's managing. As far as I can tell those labels are not persisted anywhere, so this should work as a temporary and transparent storage to pass information from the registry implementation to the provider implementation.
3. Try to get the name of the managed record in the provider implementation by calling the `toEndpointName` function from the TXT registry. I think this would kind-of break the interface between the registry and the provider.

With any of these options we could improve the grouping in the `batchChangeSet` function and then also replace my newly added snipped (which does the grouping when submitting single changes) with a call to this function.

szuecs commented:
/remove-lifecycle stale

szuecs commented:
@devkid do you mind rebasing, please?
@linki @njuettner or @Raffo will review.

devkid commented:
@szuecs done.

cw-sakamoto commented:
@devkid 
Thank you for the wonderful PR! I would like to use this feature, is it possible to resolve conflicts?

szuecs commented:
/remove-lifecycle rotten

devkid commented:
Weird, Github told me to rebase, but my local git was able to auto-merge on rebase. Conflicts are fixed now. @szuecs

szuecs commented:
/assign @linki

szuecs commented:
@devkid can you check the GH actions errors please?
```
level=error msg="Running error: interfacer: failed prerequisites: [buildssa@sigs.k8s.io/external-dns/provider/aws [sigs.k8s.io/external-dns/provider/aws.test]: analysis skipped: errors in package: [/home/runner/work/external-dns/external-dns/provider/aws/aws_test.go:735:44: undeclared name: NewDomainFilter /home/runner/work/external-dns/external-dns/provider/aws/aws_test.go:735:106: undeclared name: NewZoneIDFilter /home/runner/work/external-dns/external-dns/provider/aws/aws_test.go:735:135: undeclared name: NewZoneTypeFilter]]"
Makefile:34: recipe for target 'lint' failed
make: *** [lint] Error 3
##[error]Process completed with exit code 2.
```

devkid commented:
@szuecs linter and unit test problems are fixed.

seanmalloy commented:
/kind feature

devkid commented:
The issue with TXT records getting out of sync when a prefix/suffix is used is now fixed by adding a `ownedRecord` label to TXT endpoints. Route 53 transactions will now be grouped by ownership relation as well as by name, so this should not happen anymore.

devkid commented:
/remove-lifecycle rotten

Raffo commented on provider/aws/aws.go:
```diff
@@ -494,9 +497,30 @@ func (p *AWSProvider) submitChanges(ctx context.Context, changes []*route53.Chan
+		var newChanges, retriedChanges []*route53.Change
+		for _, c := range cs {
+			found := false
+			if p.failedChangesQueue[z] != nil {
```
I think this check is useless, if `p.failedChangesQueue[z]` the range will never run. Please remove it.

devkid replied:
The range simply goes over _all_ desired changes (no matter if they have failed before - because the loop is finding out _if_ they have failed before and sort them into new changes / changes that have failed before). Therefor, `p.failedChangesQueue[z]` is actually most likely to be `nil`.

Raffo replied:
I understand, but that's no problem for the range that you are doing the line below, see:

```
package main

import (
	"fmt"
)

func main() {
	var nilMap map[string]string
	nilMap = nil // making this more explicit that it needs to be :-)
	for _, val := range nilMap{
		fmt.Println(val)
	}
}
```

devkid replied:
I see, didn't know that `for` works on a `nil` map :upside_down_face: Will remove the check.

szuecs replied:
@alfredkrohmer as far as I see this was not done.

linki replied:
@alfredkrohmer I agree, let's remove the `if`. Should be fine.
@szuecs There should be no need for a lock since it all runs sequentially.

Raffo commented on provider/aws/aws.go:
```diff
@@ -494,9 +497,30 @@ func (p *AWSProvider) submitChanges(ctx context.Context, changes []*route53.Chan
+		for _, c := range cs {
+			found := false
+			if p.failedChangesQueue[z] != nil {
+				for _, failedChange := range p.failedChangesQueue[z] {
+					if c == failedChange {
+						retriedChanges = append(retriedChanges, c)
+						found = true
+					}
+				}
+				p.failedChangesQueue[z] = nil // clear the queue
```
Why are we doing this loop if we are going to append them anyway in line 517?

devkid replied:
The purpose of the loop is to sort them changes into new changes / changes that have failed last time. We clear the list here because the change might succeed this time (if it doesn't, it gets added back to the list below in the error handling). On line 517 the actual "sorting" is happening in a sense that new changes are tried first (because they have a higher likelihood of succeeding), and retried changes are coming after the new changes.

Raffo replied:
Do I understand it right that this has no impact on correctness? Does this likelihood have any sort of impact on actual usage? If not, can we remove it and simplify the code?

devkid replied:
As far as I can tell it has no influence on the correctness, however it reduces the overall amount of API calls (which can help prevent API rate limiting) because new changes have a higher likelihood of succeeding - they are tried first in a separate transaction, which is unlikely to fail. If they would be randomly mixed with previously failing changes, on *every* new change, you would have additional API calls (complete batch with new+old changes fails, then all changes are tried one by one). So if you have e.g. 4 failing changes and 4 new ones, you would do 9 calls (on batch with all 8 changes that fails + one batch with each change for retry) instead of only 5 (one batch with the 4 new changes + one batch with each failing change for retry).

linki replied:
I like the idea of trying to handle the new changes in their own transaction first (pretty much mimicking the "happy path" as if we didn't have any failed changes in the queue) because there are likely to succeed. The users of these records wouldn't be impacted by any delay due to the handling of failed batches (due to splitting and more api calls).

linki replied:
@alfredkrohmer How much more complicated is the code because of that in your opinion?

linki replied:
Oh I see, it's basically just this part here where we separate the `failedChanges` from the `desiredChanges` in order to move the remaining (`newChanges`) to the front of the queue.

If that's true we could make this more obvious for the reader with comments for some meaningful function names.

Raffo commented on provider/aws/aws.go:
```diff
@@ -505,17 +529,51 @@ func (p *AWSProvider) submitChanges(ctx context.Context, changes []*route53.Chan
-					log.Error(err) //TODO(ideahitme): consider changing the interface in cases when this error might be a concern for other components
-					failedUpdate = true
+					log.Errorf("Failure in zone %s [Id: %s] when submitting change batch", aws.StringValue(zones[z].Name), z)
+					log.Error(err)
```
Let's take the opportunity of this PR to make this log line better and add a message instead of just printing the error.

devkid replied:
Fixed (hopefully, wasn't exactly sure what you meant).

Raffo commented on provider/aws/aws.go:
```diff
@@ -505,17 +529,51 @@ func (p *AWSProvider) submitChanges(ctx context.Context, changes []*route53.Chan
+					changesByOwnership := groupChangesByNameAndOwnership(b)
+
+					if len(changesByOwnership) > 1 {
+						log.Error("Trying to submit change sets one-by-one instead")
```
This log should be debug, not error.

devkid replied:
Fixed

Raffo commented on provider/aws/aws.go:
```diff
@@ -505,17 +529,51 @@ func (p *AWSProvider) submitChanges(ctx context.Context, changes []*route53.Chan
+
+						for _, changes := range changesByOwnership {
+							for _, c := range changes {
+								log.Infof("Desired change: %s %s %s [Id: %s]", *c.Action, *c.ResourceRecordSet.Name, *c.ResourceRecordSet.Type, z)
```
This should probably be debug level.

devkid replied:
Fixed

Raffo commented on provider/aws/aws.go:
```diff
@@ -505,17 +529,51 @@ func (p *AWSProvider) submitChanges(ctx context.Context, changes []*route53.Chan
+									p.failedChangesQueue[z] = append(p.failedChangesQueue[z], changes...)
+								}
+							} else {
+								log.Info("Change(s) successful")
```
This log line should be removed entirely.

devkid replied:
Fixed

Raffo commented on provider/aws/aws.go:
```diff
@@ -653,9 +711,65 @@ func (p *AWSProvider) newChange(action string, ep *endpoint.Endpoint, recordsCac
 	return change, dualstack
 }
 
+// HACK: we have no easy way of passing the "ownedRecord" label of endpoints to `submitChanges` for grouping
+// transactions by owner relationship; we cheat here by passing the label as the `AliasTarget` of the record;
+// the `AliasTarget` setting is recognized and removed in `submitChanges` before submission of the change to Route53;
+// the "ownedRecord" label should only ever appear on TXT endpoints created by the TXT registry, so it should not
+// interfere with any other record that could possibly have an actual `AliasTarget` setting
+func setOwnedRecord(change *route53.Change, ep *endpoint.Endpoint) {
+	if ownedRecord, ok := ep.Labels[endpoint.OwnedRecordLabelKey]; ok && change.ResourceRecordSet.AliasTarget == nil {
+		change.ResourceRecordSet.AliasTarget = &route53.AliasTarget{
+			HostedZoneId: aws.String(endpoint.OwnedRecordLabelKey),
+			DNSName:      aws.String(ownedRecord),
+		}
+	}
+}
```
What alternatives do we have to avoid this hack?

devkid replied:
The only alternative I was seeing was to wrap the `route53.Change` type in order to add a new field, but this would require pulling the change through _all_ methods and tests that handle the changes. Also, the `ownedRecord` field would semantically belong more to the record struct than to the change struct, but we can't wrap the record struct because it's nested in the change struct, so this would probably cause even more confusion.

devkid replied:
Please let me know what you prefer.

Raffo replied:
I am not sure either, would it be too much effort to see how the `route53.Change` wrapping looks like? I don't think we should merge this hack as it is and that we must find a better way.

devkid replied:
I'll try to prepare something for this.

Raffo replied:
@devkid any update on this one?

alfredkrohmer replied:
@Raffo rebased and did the changes to wrap `route53.Change`. Let me know what you think of this.

Sorry for the _very_ late response here, I was lacking the time and motivation to continue this, both should be fixed now :)

Raffo reviewed (commented):
@devkid Apologies for the long time this took, I added a few comments and I think there are a few things that need to be addressed before we can get this in. Please do not force push to your branch again as this makes following up on PRs much harder. I will keep this open so that we can have a decent turnaround and make sure we can close the work here soon.

szuecs commented:
/remove-lifecycle rotten

szuecs commented:
/important-soon

szuecs commented:
/important soon

szuecs commented:
/remove-lifecycle stale

szuecs commented:
/remove-lifecycle stale

alfredkrohmer commented:
/remove-lifecycle rotten

linki commented:
@alfredkrohmer Thanks for the PR.

I tested it against our e2e test suite over at: https://github.com/zalando-incubator/kubernetes-on-aws/pull/5604

First by permanently introducing a broken record such that ExternalDNS gets completely stuck and doesn't do any updates. It consequently causes our e2e tests to fail because some parts rely on ExternalDNS managing records.

I then used this version (rebased on the `v0.13.1` tag) to validate that it fixes the problem. It does successfully ignore the broken record and processes the remaining records as usual.

@Raffo @szuecs How can we help to get this merged?

szuecs commented on provider/aws/aws.go:
```diff
@@ -520,17 +560,49 @@ func (p *AWSProvider) submitChanges(ctx context.Context, changes []*route53.Chan
+
+					changesByOwnership := groupChangesByNameAndOwnershipRelation(b)
+
+					if len(changesByOwnership) > 1 {
```
I am not sure why this condition is `>1` and a change with `==1` would be a failedUpdate

alfredkrohmer replied:
If `len(changesByOwnership) == 1`, then the below code would just repeat the same single change batch, which would most likely just fail again.

szuecs commented on provider/aws/aws_test.go:
```diff
@@ -807,22 +827,95 @@ func TestAWSsubmitChangesError(t *testing.T) {
+	for _, r := range records {
+		if r.DNSName == "success.zone-1.ext-dns-test-2.teapot.zalan.do" {
+			ok = true
+		} else if r.DNSName == "fail.zone-1.ext-dns-test-2.teapot.zalan.do" || r.DNSName == "fail__edns_housekeeping.zone-1.ext-dns-test-2.teapot.zalan.do" {
```
In general I think here would be a `switch r.DNSName {}` more readable.

alfredkrohmer replied:
Changed

szuecs commented on provider/aws/aws_test.go:
```diff
@@ -807,22 +827,95 @@ func TestAWSsubmitChangesError(t *testing.T) {
+	require.NoError(t, err)
+	count := 0
+	for _, r := range records {
+		if r.DNSName == "success.zone-1.ext-dns-test-2.teapot.zalan.do" || r.DNSName == "fail.zone-1.ext-dns-test-2.teapot.zalan.do" || r.DNSName == "success2.zone-1.ext-dns-test-2.teapot.zalan.do" || r.DNSName == "fail__edns_housekeeping.zone-1.ext-dns-test-2.teapot.zalan.do" {
```
same here -> use `switch`

alfredkrohmer replied:
Changed

szuecs commented:
/assign @alfredkrohmer

szuecs commented:
/approve

szuecs commented:
/assign @njuettner

linki commented on provider/aws/aws.go:
```diff
@@ -337,7 +356,7 @@ func (p *AWSProvider) records(ctx context.Context, zones map[string]*route53.Hos
 			}
 
-			if r.AliasTarget != nil {
+			if r.AliasTarget != nil && aws.StringValue(r.AliasTarget.HostedZoneId) != endpoint.OwnedRecordLabelKey { // see HACK in `newChanges`
```
What does this do? Who does the `HostedZoneId` has anything to do with `OwnedRecordLabelKey` itself (`ownedRecord`)?

Also there's no `HACK` anywhere in the file. Is this still needed?

alfredkrohmer replied:
Sorry, this was a left-over from the previous implementation approach. Removed.

linki commented on provider/aws/aws.go:
```diff
@@ -509,9 +528,30 @@ func (p *AWSProvider) submitChanges(ctx context.Context, changes []*route53.Chan
+				for _, failedChange := range p.failedChangesQueue[z] {
+					if c == failedChange {
+						retriedChanges = append(retriedChanges, c)
+						found = true
```
Do we maybe want to add a `break` here? It seems to be a common pattern. If `c` was found in `failedChangesQueue` we don't need to keep looking. Also, if for some reason it's found multiple times in `failedChangesQueue` it would be added to `retriedChanges` multiple times as well. Not sure if that's what you want or not, though.

linki replied:
It would be helpful to extract these lines into a function `findChangeInQueue(c, p.failedChangesQueue[z])` that takes the item `c` and the queue where you look for it. That's essentially all that's happening on lines 534 to 541.

alfredkrohmer replied:
Moved out the whole chunk into `findChangesInQueue`.

linki commented on provider/aws/aws.go:
```diff
@@ -509,9 +528,30 @@ func (p *AWSProvider) submitChanges(ctx context.Context, changes []*route53.Chan
 		for i, b := range batchCs {
+			if len(b) == 0 {
+				continue
+			}
```
How come we need this now? Can `b` now be empty where it wasn't possible before? If the change above is just recording the `newChanges` and `failedChanges` the batch size should be the same as before.

linki replied:
Ok, I think I know why we need it. `retriedChanges` can obviously be empty and so can `newChanges` if all desired records were part of `failedChangesQueue`, i.e. have been tried before.

linki commented on provider/aws/aws.go:
```diff
@@ -520,17 +560,49 @@ func (p *AWSProvider) submitChanges(ctx context.Context, changes []*route53.Chan
+								if _, ok := p.failedChangesQueue[z]; !ok {
+									p.failedChangesQueue[z] = changes
+								} else {
+									p.failedChangesQueue[z] = append(p.failedChangesQueue[z], changes...)
```
You should be able to remove the `if` here. if `p.failedChangesQueue[z] == nil` then the append will become `append(nil, "foo")` which works just fine.

alfredkrohmer replied:
Done

linki commented on provider/aws/aws.go:
```diff
@@ -687,9 +761,28 @@ func (p *AWSProvider) newChange(action string, ep *endpoint.Endpoint) (*route53.
 		change.ResourceRecordSet.HealthCheckId = aws.String(prop.Value)
 	}
 
+	if ownedRecord, ok := ep.Labels[endpoint.OwnedRecordLabelKey]; ok {
```
I get the point of that label. However, the `Labels` are intended to be persisted in the TXT records such that ExternalDNS can access that information across restarts or if multiple ExternalDNS' work within the same Hosted Zone.

However, I don't see that this persistence is needed and that this label is only required to hold this information while the Endpoint is processed within ExternalDNS during a single synchronisation run (comparable to `failedChangesQueue` which gets lost when ExternalDNS stops).

For this particular need (where persistence isn't needed) one should use the map `endpoint.ProviderSpecific` instead of `endpoint.Labels` because these are not persisted.

alfredkrohmer replied:
These `Labels` are only added to the `TXT` records themselves; the labels of these are not persisted (it's the other way around, the TXT records are persisting the labels of the records they "own"). I would disagree that this particular field should be in `ProviderSpecific` instead as this ownership relation itself is not provider-specific in any way and could be picked up by the `ApplyChanges` implementation of any other provider, too.

linki replied:
Ok, I see. If it's only on the TXT records and not persisted (because TXT records don't have TXT records for themselves to store them) then I don't mind.

linki commented on provider/aws/aws_test.go:
```diff
@@ -807,22 +827,101 @@ func TestAWSsubmitChangesError(t *testing.T) {
+		},
+	}
+	clientStub.MockMethod("ChangeResourceRecordSets", input1).Return(nil, fmt.Errorf("Mock route53 failure"))
+
```
Shouldn't we run `require.Error(t, provider.submitChanges(...))` here?

alfredkrohmer replied:
No, `cs1` / `input1` and `cs2` / `input2` are mocks for the invocation of `ChangeResourceRecordSets` that happen in the single invocation `submitChanges` below. The first invocation of `ChangeResourceRecordSets` is attempting to create the succeeding and failing record in one batch, the second invocation is retrying the failing record.

linki commented on provider/aws/aws_test.go:
```diff
@@ -807,22 +827,101 @@ func TestAWSsubmitChangesError(t *testing.T) {
+			"fail.zone-1.ext-dns-test-2.teapot.zalan.do",
+			"success2.zone-1.ext-dns-test-2.teapot.zalan.do",
+			"fail__edns_housekeeping.zone-1.ext-dns-test-2.teapot.zalan.do":
+			count += 1
```
I think it would be better to have a function that checks whether a particular record, given its name, is part of a list of records.

This way we could check the presence (and absence for above) more explicitly and avoid the obscure `count` and `ok` variables, e.g.

```go
// check inclusion
require.True(t, containsRecordWithDNSName("success.zone-1.ext-dns-test-2.teapot.zalan.do", records)
// check exclusion
require.False(t, containsRecordWithDNSName("fail.zone-1.ext-dns-test-2.teapot.zalan.do", records)
```

linki commented on provider/aws/aws_test.go:
```diff
@@ -836,27 +935,33 @@ func TestAWSBatchChangeSet(t *testing.T) {
-				Name: aws.String(fmt.Sprintf("host-%d", i)),
-				Type: aws.String("A"),
+		cs = append(cs,
+			&Route53Change{
```
There's a lot of places in this PR where the only change is that a `route53.Change` is wrapped by our custom `Route53Change{Change: x}`. If we could separate this change from the actually meaningful part (handling error and retry) then the PR would get much smaller.

We could isolate this change in a single commit and the rest in another commit. Or we just create another PR where we only do the `Route53Change` wrapping (without changing any behaviour) and rebase this PR onto it.

Raffo replied:
I agree it would be a good idea to do the replacement as @linki is suggesting.

alfredkrohmer replied:
Done.

linki commented on registry/txt.go:
```diff
@@ -194,11 +194,13 @@ func (im *TXTRegistry) generateTXTRecord(r *endpoint.Endpoint) []*endpoint.Endpo
 	// old TXT record format
 	txt := endpoint.NewEndpoint(im.mapper.toTXTName(r.DNSName), endpoint.RecordTypeTXT, r.Labels.Serialize(true)).WithSetIdentifier(r.SetIdentifier)
 	txt.ProviderSpecific = r.ProviderSpecific
+	txt.Labels[endpoint.OwnedRecordLabelKey] = r.DNSName
```
This looks a bit misplaced. The TXT registry is shared by all providers, yet `endpoint.OwnedRecordLabelKey` is only used in the AWS provider. I can imagine that this could apply to all other providers as well but maybe it's better to make it part of `r.ProviderSpecific` (one line before) such that we don't need this line.

Raffo commented:
@alfredkrohmer would you mind addressing @linki's comments. We would love to get this finally merged as it would be addressing some longstanding issues with the AWS provider.

alfredkrohmer commented:
Rebased and split into two commits for better reviewability. Only open discussion is around `Labels` vs `ProviderSpecific` (see my comment above in the thread).

linki commented on registry/txt_test.go:
```diff
@@ -867,11 +867,17 @@ func testTXTRegistryMissingRecordsNoPrefix(t *testing.T) {
 			// owner is taken from the source record (A, CNAME, etc.)
 			Targets:    endpoint.Targets{"\"heritage=external-dns,external-dns/owner=owner\""},
 			RecordType: endpoint.RecordTypeTXT,
+			Labels: endpoint.Labels{
```
See comment from @alfredkrohmer above for the resolution on this.

linki commented:
@Raffo @szuecs @njuettner LGTM

@alfredkrohmer Thanks a lot for persistently working on this! 🙏  It will solve a long-standing issue with ExternalDNS on AWS.

linki commented:
/lgtm

linki commented:
@Raffo Would you mind cutting a new release with this included?

Raffo commented:
@linki i did a release last weekend. I'm totally up to do a release, but maybe let's aim for the beginning of February so that we have a few more changes. Wdyt?

linki commented:
The sooner the better for me. I tested this fix through our pipeline but I would like to run the official version going forward: https://github.com/zalando-incubator/kubernetes-on-aws/pull/5604

Raffo commented:
@linki sounds good. I'll try to get some more fixes in and then get a release cooking at the beginning of February.
